src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -54,14 +65,106 @@ public Task<bool> IsEnabledAsync<TContext>(string feature, TContext appContext,
+        }
+
+        private async ValueTask<T> GetVariantAsync<T, TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
+        {
+            if (feature == null)
+            {
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            FeatureVariant variant = null;
+
+            FeatureDefinition featureDefinition = await _featureDefinitionProvider.GetFeatureDefinitionAsync(feature, cancellationToken).ConfigureAwait(false);
+
+            if (featureDefinition != null)
+            {
+                if (string.IsNullOrEmpty(featureDefinition.Assigner))
+                {
+                    throw new FeatureManagementException(
+                        FeatureManagementError.InvalidConfiguration,
+                        $"Invalid feature variant assigner name for the feature {feature}");
+                }
+
+                IFeatureVariantAssignerMetadata assigner = GetFeatureAssignerMetadata(featureDefinition.Assigner);
+
+                if (assigner == null)
+                {
+                    string errorMessage = $"The feature assigner '{featureDefinition.Assigner}' specified for feature '{feature}' was not found.";
+
+                    if (!_options.IgnoreMissingFeatureAssigners)
+                    {
+                        throw new FeatureManagementException(FeatureManagementError.MissingFeatureAssigner, errorMessage);
+                    }
+                    else
+                    {
+                        _logger.LogWarning(errorMessage);
+
+                        return default(T);
+                    }
+                }
+
+                var context = new FeatureVariantAssignmentContext()
+                {
+                    FeatureDefinition = featureDefinition
+                };
+
+                //
+                // IContextualFeatureVariantAssigner
+                if (useAppContext)
+                {
+                    ContextualFeatureVariantAssignerEvaluator contextualAssigner = GetContextualFeatureVariantAssigner(featureDefinition.Assigner, typeof(TContext));
+
+                    if (contextualAssigner != null)
+                    {
+                        variant = await contextualAssigner.AssignVariantAsync(context, appContext, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+
+                //
+                // IFeatureVariantAssigner
+                if (assigner is IFeatureVariantAssigner featureVariantAssigner)
+                {
+                    variant = await featureVariantAssigner.AssignVariantAsync(context, cancellationToken).ConfigureAwait(false);
+                }
+            
+                if (variant == null &&
+                    featureDefinition.Variants != null)
+                {
+                    variant = featureDefinition.Variants.FirstOrDefault(v => v.Default);
```

Comment:
I am wondering if we should fail if the user didn't specify a default variant in this case. Returning a default(T) could be a silent failure that is hard to debug.

Reply:
I think that's a good idea.

Reply:
I'm wondering if it makes sense to throw whenever we encounter a feature that doesn't have a default variant. That way, it doesn't come as a surprise in the off chance that assignment comes down to default variant.

Reply:
Sure, that will make things a lot simpler.

Reply:
Added.